test_integration/base_suite_test.go

```diff
@@ -0,0 +1,285 @@
+		tc.WithStackFiles(composeFilePath),
+	)
+	s.Require().NoErrorf(err, "Error preparing Docker Compose Stack: %q\n", err)
+
+	s.ComposeStack = composeStack
+
+	err = composeStack.Up(s.Context, tc.Wait(true), tc.WithRecreate(api.RecreateForce))
+	s.Require().NoErrorf(err, "Error creating Docker Compose Stack: %q\n", err)
+}
+
+func (s *BaseDockerComposeTestSuite) downWithDockerCompose() {
+	err := s.ComposeStack.(interface {
+		Down(context.Context, ...tc.StackDownOption) error
+	}).Down(
+		context.Background(),
+		tc.RemoveImagesLocal,
+		tc.RemoveOrphans(true),
+		tc.RemoveVolumes(true),
+	)
+	s.Require().NoErrorf(err, "Error tearing down docker compose stack: %q\n", err)
+	s.Cancel()
+}
+
+func (s *BaseDockerComposeTestSuite) executeCommandInContainer(containerName string, cmd []string) (string, error) {
+	typeAssertedComposeStack, ok := s.ComposeStack.(interface {
+		ServiceContainer(context.Context, string) (*testcontainers.DockerContainer, error)
+	})
+	if !ok {
+		return "", fmt.Errorf("executing command inside container: ComposeStack does not implement ServiceContainer method")
+	}
+
+	container, err := typeAssertedComposeStack.ServiceContainer(s.Context, containerName)
+	if err != nil {
+		return "", fmt.Errorf("executing command inside container: failed to get service container: %w", err)
+	}
+
+	exitCode, reader, err := container.Exec(s.Context, cmd)
+	if err != nil {
+		return "", fmt.Errorf("executing command inside container: failed to execute command: %w", err)
+	}
+
+	output, err2 := readContainerExecOutput(err, reader)
+	if err2 != nil {
+		return "", err2
+	}
+
+	if exitCode != 0 {
+		return output, fmt.Errorf("executing command inside container: command exited with status %d: %s", exitCode, output)
+	}
+
+	return output, nil
+}
+
+func (s *BaseDockerComposeTestSuite) executeCommandInDefaultJumpbox(cmd []string) (string, error) {
+	return s.executeCommandInContainer("bacalhau-jumpbox-node", cmd)
+}
+
+func (s *BaseDockerComposeTestSuite) waitForJobToComplete(jobID string, timeout time.Duration) (time.Duration, error) {
+	startTime := time.Now()
+	endTime := startTime.Add(timeout)
+
+	for time.Now().Before(endTime) {
+		jobDescriptionResultJson, err := s.executeCommandInDefaultJumpbox([]string{"bacalhau", "job", "describe", "--output=json", jobID})
+		if err != nil {
+			return time.Since(startTime), err
+		}
+
+		jobState, err := extractJobStateType(jobDescriptionResultJson)
+		if err != nil {
+			return time.Since(startTime), err
+		}
+
+		if jobState == "Completed" {
+			return time.Since(startTime), nil
+		}
+
+		time.Sleep(1000 * time.Millisecond)
+	}
+
+	return timeout, fmt.Errorf("job did not finish within allowed time limit %s", timeout.String())
+}
+
+func (s *BaseDockerComposeTestSuite) renderDockerComposeFile(inputFilePath string, tmpDir string, imageReplacements map[string]string) (string, error) {
+	data, err := os.ReadFile(inputFilePath)
+	if err != nil {
+		return "", fmt.Errorf("error reading file: %v", err)
```

Comment:
nit: should we instead use `%w`: https://go.dev/blog/go1.13-errors#wrapping-errors-with-w ?

Reply:
Good point. I would consider it when complex context is needed to be passed down, but to your point I should have used `%q` for clarity